File: leetcode/1036-rotting-oranges/solution.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int fresh = 0;
        
        Queue<int[]> q = new LinkedList<>();

        //if the orange is rotten, push it to the queue
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 2) q.offer(new int[]{i,j});
                else if(grid[i][j] == 1) fresh++;
            }
        }
        if(fresh == 0) return 0;

        int level = 0;
        int[][] directions = new int[][]{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        while(!q.isEmpty()){
            int size = q.size();
            while(size-- > 0){
                int[] curr = q.poll();
                for(int[] dir : directions){
                    int newX = curr[0] + dir[0], newY = curr[1] + dir[1];
                    
                    //if within bounds, and fresh, mark it rotten and add it to the queue
                    if(newX >= 0 && newX < m && newY >= 0 && newY < n && grid[newX][newY] == 1){
                        fresh--;
                        grid[newX][newY] = 2;
                        q.offer(new int[]{newX, newY});
                    }
                }
            }
            level++;
        }

        return fresh == 0 ? level - 1 : -1;

    }
}

File: leetcode/1036-rotting-oranges/solution.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        minutes = 0
        q = []
        fresh = 0

        # add all rotten oranges' coordinates to the queue
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    q.append([i, j])
                elif grid[i][j] == 1:
                    fresh += 1
            
        if fresh == 0:
            return 0
        
        while q:
            level_size = len(q)
            for _ in range(level_size):
                orange = q.pop(0)
                i, j = orange[0], orange[1]

                directions = [[-1,0], [1, 0], [0, -1], [0, 1]]
                for di, dj in directions:
                    ni, nj = i + di, j + dj
                    if 0 <= ni < m and 0 <= nj < n and grid[ni][nj] == 1:
                        fresh -= 1
                        grid[ni][nj] = 2
                        q.append([ni, nj])
            minutes += 1
         
        return -1 if fresh != 0 else minutes - 1
    
    
